modules/core/include/opencv2/core/quaternion.inl.hpp

```diff
@@ -866,6 +866,766 @@ Quat<T> Quat<T>::spline(const Quat<T> &q0, const Quat<T> &q1, const Quat<T> &q2,
+        Quat<T> q2 = Quat<T>::createFromXRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromZRot(angles(2) / 2.);
+        q = q3 * q2 * q1;
+        break;
+    }
+    case INT_XYX:
+    {
+        Quat<T> q1 = Quat<T>::createFromXRot(angles(0) / 2.);
+        Quat<T> q2= Quat<T>::createFromYRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromXRot(angles(2) / 2.);
+        q = q1 *q2 * q3;
+        break;
+    }
+    case EXT_XYX:
+    {
+        Quat<T> q1 = Quat<T>::createFromXRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromYRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromXRot(angles(2) / 2.);
+        q = q3 * q2 * q1;
+        break;
+    }
+    case INT_ZYZ:
+    {
+        Quat<T> q1 = Quat<T>::createFromZRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromYRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromZRot(angles(2) / 2.);
+        q = q1 *q2 * q3;
+        break;
+    }
+    case EXT_ZYZ:
+    {
+        Quat<T> q1 = Quat<T>::createFromZRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromYRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromZRot(angles(2) / 2.);
+        q = q3 * q2 * q1;
+        break;
+    }
+    case INT_YZY:
+    {
+        Quat<T> q1 = Quat<T>::createFromYRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromZRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromYRot(angles(2) / 2.);
+        q = q1 * q2 * q3;
+        break;
+    }
+    case EXT_YZY:
+    {
+        Quat<T> q1 = Quat<T>::createFromYRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromZRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromYRot(angles(2) / 2.);
+        q = q3 * q2 * q1;
+        break;
+    }
+    case INT_XZX:
+    {
+        Quat<T> q1 = Quat<T>::createFromXRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromZRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromXRot(angles(2) / 2.);
+        q = q1 * q2 * q3;
+        break;
+    }
+    case EXT_XZX:
+    {
+        Quat<T> q1 = Quat<T>::createFromXRot(angles(0) / 2.);
+        Quat<T> q2 = Quat<T>::createFromZRot(angles(1) / 2.);
+        Quat<T> q3 = Quat<T>::createFromXRot(angles(2) / 2.);
+
+        q = q3 * q2 * q1;
+        break;
+    }
+    default:
+        CV_Error(Error::StsBadArg, "Wrong Euler angles type!" );
+        break;
+    }
+
+    return q;
+}
+
+template <typename T>
+Vec<T, 3> Quat<T>::toEulerAngles(EulerAnglesType eulerAnglesType){
+    Matx33d R = toRotMat3x3();
+    Vec<T, 3> angles;
+
+    switch (eulerAnglesType)
+    {
+
+    case INT_XYZ:
+    {
+        if (abs(R(0, 2) - 1) < CV_QUAT_CONVERT_THRESHOLD)
+        {
+            std::cout << "WARNING: Gimbal Lock will occur. Euler angles are non-unique." << std::endl;
+            std::cout << "For intrinsic rotations, we set the third angle to zero, and for external rotation, we set the first angle to zero." << std::endl;
```

Comment:
> std::cout

Avoid that. There is `CV_LOG_WARNING()`. However messages probably should be reduced to INFO level.

Reply:
Done!